File: STA5106_Comp_Statistics/HW5/iter_plot.m
% Plot converging sequences of Simple Iteration and Newton-Raphson
% method.
% Input:
%   fun: A real function that we want to find the root.
%   dfun: Real function, 1st order derivative of fun.
%   x0: Initial value of x.
%   eps: Absolute change stopping threshold.
% Output:
%   Display the two converging sequences

% fun = inline("0.9 * sin(x) - x", "x");
% dfun = inline("0.9 * cos(x) - 1", "x");
% iter_plot(fun, dfun, pi/4, 1e-6)

function iter_plot(fun, dfun, x0, eps)
    % SSE v.s. Iteration index
    figure(1); clf;
    title("SSE of two converging sequences","Fontsize",15);
    xlabel("X");
    ylabel("Sum of Squared Error (SSE)");    
    hold on;
    [~,seq1] = simple_iter(fun, x0, eps);
    [~,seq2] = newton(fun, dfun, x0, eps);
    p1 = plot(seq1.^2,"Color","b","MarkerSize",12);
    hold on;
    p2 = plot(seq2.^2,"--","Color","r");
    legend([p1,p2],"Simple Iteration", "Newton Method");
    % Compare the converging sequences on function graph directly
    figure(2); clf;
    t = 0:0.01:1;
    plot(t,fun(t));
    title("Converging sequences of two methods","Fontsize",15);
    xlabel("X");
    ylabel("fun");    
    hold on;
    [~,seq1] = simple_iter(fun, x0, eps);
    [~,seq2] = newton(fun, dfun, x0, eps);
    p1 = plot(seq1, fun(seq1),".","Color","b","MarkerSize",12);
    hold on;
    p2 = plot(seq2, fun(seq2),"*","Color","r");
    legend([p1,p2],"Simple Iteration", "Newton Method");
end

File: STA5106_Comp_Statistics/HW5/iter_plot.py
import numpy as np
import matplotlib.pyplot as plt

def simple_iter(fun, x0, eps):
    x = x0;
    seq = [x0];
    while True:
        x1 = x + fun(x)
        seq.append(x1)
        if abs(x1 - x) < eps:
            break
        x = x1;
    return x1, seq

def newton(fun, dfun, x0, eps):
    x = x0;
    seq = [x0];
    while True:
        x1 = x - fun(x) / dfun(x);
        seq.append(x1)
        if abs(x1 - x) < eps:
            break
        x = x1;
    return x1, seq

def iter_plot(fun, dfun, x0, eps):
    [_, seq1] = simple_iter(fun, x0, eps)
    [_, seq2] = newton(fun, dfun, x0, eps)
    plt.figure(1)
    plt.plot(seq1, linestyle="-", color="blue",label="Simple Iteration")
    plt.plot(seq2, linestyle="--", color="red",label="Newton Method")
    plt.xlabel("X")
    plt.ylabel("Sum of Squared Error (SSE)")
    plt.title("SSE of two Converging Sequences",fontsize=15)
    plt.legend(loc="upper right")
    plt.show()
    plt.figure(2)
    t = list(np.arange(0,1,0.01))
    plt.plot(t, fun(t), color="black", linestyle="-")
    plt.scatter(seq1, fun(seq1), marker=".", color="blue",label="Simple Iteration")
    plt.scatter(seq2, fun(seq2), marker="*", color="red",label="Newton Method")
    plt.xlabel("X")
    plt.ylabel("fun")
    plt.title("Converging sequences of two methods",fontsize=15)
    plt.legend(loc="upper right")
    plt.show()

fun = lambda x: 0.9 * np.sin(x) - x
dfun = lambda x: 0.9 * np.cos(x) - 1
iter_plot(fun, dfun, np.pi/4, 1e-6)
